### PR TITLE
Implement pattern types as the backing logic of rustc_scalar_valid_range attributes

### PR DESCRIPTION
**Pull request by oli-obk**
Alternative to https://github.com/rust-lang/rust/pull/103724

Some explanation of the three alternatives can be found [on my blog](https://cohost.org/oli-obk/post/203456-strong-typing-has-do)

r? @BoxyUwU 

cc @joshtriplett 

This PR causes fields of types that use `rustc_scalar_valid_range` to be of a new `TyKind`: the `Pat` kind which at present has a nested type and a u128 range of valid values. This represents the status quo exactly, but before I rip out the old attribute checks and implement them on top of pattern types, I wanted to open this PR for discussion.

For now I added an inherently unsafe way to construct these types: you cast an arbitrary other type to them. So initializing `NonNull` works similar to `NonNull { pointer: some_raw_ptr as _ }`. The underscore figures our the right type, as right now there is *no user visible syntax* for these types.

As a next step I would like to remove the cast and instead implement it so that initialization of `NonNull` works like

```rust
match some_value {
    value @ 1.. => Some(NonZeroU8 { value }),
    _ => None,
}
```

where we teach pattern matching that the type of `pointer` is one of the new pattern types. The type of `value` is `(*const T) is 1..`

At this stage we would not support

```rust
match some_value {
    0 => None,
    value => Some(NonZeroU8 { value }),
}
```

because that requires reasoning across arms, which requires more work than just doing some local reasoning. The compiler *already* has this information for exhaustiveness checking, so we're not inventing anything new here, but we need to still fetch the information from exhaustiveness checking, so it seems best to start simple.

### CODE REVIEW

File: library/std/src/sys/unix/time.rs [hunk 1]
```diff
@@ -9,12 +9,24 @@ pub const UNIX_EPOCH: SystemTime = SystemTime { t: Timespec::zero() };
 pub const TIMESPEC_MAX: libc::timespec =
     libc::timespec { tv_sec: <libc::time_t>::MAX, tv_nsec: 1_000_000_000 - 1 };
 
-#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[derive(Copy, Clone, Eq, Ord, Hash)]
```
Comment (bjorn3) [inline library/std/src/sys/unix/time.rs:12]: This will lose `StructuralPartialEq` for `SystemTime`, right?

Comment (oli-obk) [reply library/std/src/sys/unix/time.rs:12]: oops yea, I'll fix that.

**Comment by oli-obk** [comment]
Hmm... I cannot reproduce the CI failure locally, even with llvm built locally and with debug assertions on. Seems to be a LLVM 13 bug

@bors try @rust-timer queue

File: src/rustdoc-json-types/lib.rs [hunk 2]
```diff
@@ -565,6 +565,12 @@ pub enum Type {
         type_: Box<Type>,
         len: String,
     },
+    /// u32 is 0..=100
+    Pat {
```
Comment (aDotInTheVoid) [inline src/rustdoc-json-types/lib.rs:569]: Could you bump the `FORMAT_VERSION` constant in the top of this file please.

Also, CI will fail unless you add a new case to

https://github.com/rust-lang/rust/blob/226b2496fc3ef997fa1cccf84259539203f82375/src/tools/jsondoclint/src/validator.rs#L243-L245

All it needs to do is recurse into the inner type.

File: compiler/rustc_const_eval/src/interpret/visitor.rs [hunk 3]
```diff
@@ -131,6 +137,15 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> Value<'mir, 'tcx, M> for OpTy<'tc
     ) -> InterpResult<'tcx, Self> {
         ecx.operand_field(self, field)
     }
+
+    #[inline(always)]
+    fn strip_pat_ty(mut self) -> Self {
+        let ty::Pat(raw_ptr_ty, _) = *self.layout.ty.kind() else {
+            bug!("NonNull must contain a pattern type, but had {}", self.layout.ty)
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/visitor.rs:144]: These assertions are confusing. The method name and doc comment to not talk about NonNull in any way.

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs [hunk 4]
```diff
@@ -83,6 +83,9 @@ pub(crate) fn eval_nullary_intrinsic<'tcx>(
             ty::Alias(..) | ty::Param(_) | ty::Placeholder(_) | ty::Infer(_) => {
                 throw_inval!(TooGeneric)
             }
+            ty::Pat(..) => {
+                unimplemented!("pattern types need to calculate pattern from their pattern")
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/intrinsics.rs:87]: I can't parse this panic message... too many patterns?^^

File: compiler/rustc_const_eval/src/interpret/validity.rs [hunk 5]
```diff
@@ -491,7 +491,27 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
...
+                match (*scalar, value.layout.abi) {
+                    (Immediate::Scalar(scalar), Abi::Scalar(s))
+                    | (Immediate::ScalarPair(scalar, _), Abi::ScalarPair(s, _)) => {
+                        self.visit_scalar(scalar, s)?
+                    }
+                    other => span_bug!(
+                        self.ecx.cur_span(),
+                        "invalid abi {other:?} for pattern type {ty:?}"
+                    ),
+                }
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/validity.rs:512]: Isn't this duplicating the scalar range check after this `match`?

I assume for now patterns are restricted to things that can be actually represented in the `Scalar` ABI? If `0 | 5 | 10` is accepted as a pattern then we should properly check it here...

**Comment by oli-obk** [comment]
Closing in favor of https://github.com/rust-lang/rust/pull/107606

**Pull request closed without merging**